### PR TITLE
Rethinking sampling

### PR DESCRIPTION
**Pull request by josephdviviano**
This PR is a hodgepodge of a few tweaks, bugfixes, and investigations related to the sampling logic, including a new simple continuous example.

- `estimator_outputs` are now re-used when sampling off policy.
  - This is accomplished currently using padding. In a future PR this will be handled using nested tensors. I will likely wait until their API solidifies (it will apparently change in the near future). 
- `policy_kwargs` are passed around properly to do off policy sampling, hopefully in a way that remains generic.
- `TODO:s` added around based on my observations.
- `log_reward_min_clamp` now off by default and only defined in the gflownets, NOT the environment.
- Other minor tweaks I made while chasing a ghost relating to magically changing numbers (I now believe this to be due to faulty RAM on my laptop).
- API change: the user must now specify often whether a gflownet or sampler is running `off_policy` or not. This is for efficiency. If sampling happens off policy, we save estimator outputs (because we assume we will need them later, to evaluate log probabilities of actions under the policy). If it's done `on_policy`, we calculate the log-rewards during the forward pass.

### CODE REVIEW

**Comment by saleml** [comment]
I'll wait for the other PR to merge so that diffs are accurate here

**Comment by josephdviviano** [comment]
Sorry in advance for the large PR - feel free to be critical ...

**Comment by josephdviviano** [comment]
I'll wait for @saleml (who is defending next week so is likely distracted at the minute) to merge. No rush!

**Comment by saleml** [comment]
On it !

File: src/gfn/gflownet/base.py
Comment (saleml) [inline src/gfn/gflownet/base.py:217]: Where is `self.log_reward_clip_min` defined ?

**Review by saleml — COMMENTED** [review]
Sorry for my very late review. 

This is a great PR, that would make using the library much simpler. Thanks a lot @josephdviviano.

I left a few comments, questions and suggestions. They are minor. Hopefully the tests would pass after the fixes

File: src/gfn/containers/trajectories.py
Comment (josephdviviano) [reply src/gfn/containers/trajectories.py:172]: What do you think of simply:

```
        if is_tensor(self.estimator_outputs):
            estimator_outputs = self.estimator_outputs[..., index]
            estimator_outputs = estimator_outputs[:new_max_length]
```
?

File: src/gfn/containers/trajectories.py
Comment (josephdviviano) [reply src/gfn/containers/trajectories.py:245]: I think the idea is to be able to extend an empty `Trajectories` instance, say with a stored buffer.

I agree it is dangerous but I think we should support this behaviour.

Admittedly it has been some time since I looked at this so I might be forgetting something.

File: src/gfn/gflownet/base.py
Comment (josephdviviano) [reply src/gfn/gflownet/base.py:51]: Yes I suppose so. I can move `pf_pb_named_parameters` and `pf_pb_parameters`, confirming here that I should not move `logz_named_parameters` and `logz_parameters`?

File: src/gfn/gflownet/base.py
Comment (josephdviviano) [reply src/gfn/gflownet/base.py:80]: No i think this leads to dangerous unintentional user errors (it did for me).

File: src/gfn/gflownet/base.py
Comment (josephdviviano) [reply src/gfn/gflownet/base.py:87]: Because you might want to train off policy, but then sample later on policy, was my thinking.

It's easier for the user to reason about what they're doing if, when they call the sampler, they explicitly ask for on/off policy behaviour, instead of needing to check which "mode" the GFN is in.

File: src/gfn/samplers.py
Comment (josephdviviano) [reply src/gfn/samplers.py:91]: Again, I'm worried about the effect of this being implicit.

File: src/gfn/gflownet/base.py
Comment (josephdviviano) [reply src/gfn/gflownet/base.py:217]: nice catch thanks --

I guess it would be cleaner to move this to the base class. I'll leave a TODO for now.

File: src/gfn/gflownet/detailed_balance.py
Comment (josephdviviano) [reply src/gfn/gflownet/detailed_balance.py:71]: good catch! jeeze

File: src/gfn/gflownet/trajectory_balance.py
Comment (josephdviviano) [reply src/gfn/gflownet/trajectory_balance.py:35]: i dont think any of them should have a default (maybe we should have a quick call about this design decision!)

File: src/gfn/gflownet/detailed_balance.py
Comment (josephdviviano) [reply src/gfn/gflownet/detailed_balance.py:75]: Good idea. I filed an issue.

File: src/gfn/gflownet/detailed_balance.py
Comment (josephdviviano) [reply src/gfn/gflownet/detailed_balance.py:166]: AHH!

File: src/gfn/gym/discrete_ebm.py
Comment (josephdviviano) [reply src/gfn/gym/discrete_ebm.py:119]: Removed!

File: src/gfn/preprocessors.py
Comment (josephdviviano) [reply src/gfn/preprocessors.py:36]: It means that, the `identity` preprocessor is typecasting data, which seems like unexpected behaviour possibly. I would expect this to return whatever tensor is already inside `states` untouched.

**Comment by josephdviviano** [comment]
Hey Salem -- the indexing changes I implemented to fix this

https://github.com/GFNOrg/torchgfn/pull/147#discussion_r1430087521

have broken the tests -- I'm working on that now. I opened a can of worms here! There's likely an elegant solution.

**Comment by josephdviviano** [comment]
Finally I revered the change for that failing test. I'm not sure there's a better solution that isn't extremely complicated.

I'd like to get this PR merged but we can happily revisit this issue in a future much smaller PR.

**Comment by josephdviviano** [comment]
@saleml would love you to check this before I merge :)

File: src/gfn/containers/trajectories.py
Comment (saleml) [reply src/gfn/containers/trajectories.py:180]: that should work !

File: src/gfn/containers/trajectories.py
Comment (saleml) [reply src/gfn/containers/trajectories.py:257]: Fair enough!

**Comment by saleml** [comment]
This is some great work! Thank you Joseph for this very important PR.

I have read your replies to my comments, and seen that the tests pass. I think this can be merged as is.

Small nitpick: For this comment: https://github.com/GFNOrg/torchgfn/pull/147#discussion_r1430099166, do you think we should add the argument in the abstract function?

**Review by saleml — APPROVED** [review]

**Comment by josephdviviano** [comment]
> Small nitpick: For this comment: https://github.com/GFNOrg/torchgfn/pull/147#discussion_r1430099166, do you think we should add the argument in the abstract function?

I added it!

**Pull request merged**
